src/pip/_internal/wheel_builder.py

```diff
@@ -223,15 +222,18 @@ def _build_one_inside_env(
                 source_dir=req.unpacked_source_directory,
                 global_options=global_options,
                 build_options=build_options,
-                tempd=temp_dir.path,
+                wheel_directory=wheel_directory,
             )
 
         if wheel_path is not None:
             wheel_name = os.path.basename(wheel_path)
             dest_path = os.path.join(output_dir, wheel_name)
             try:
                 wheel_hash, length = hash_file(wheel_path)
-                shutil.move(wheel_path, dest_path)
+                # We can do a rename here because wheel_path is guaranteed to be
+                # in the same filesystem as output_dir. An atomic is rename
+                # to avoid concurrency issues when populating the cache.
+                os.rename(wheel_path, dest_path)
```

Comment:
Why switch from `shutil.move` to `os.rename`? `shutil.move` uses rename first and falls back to copy: https://docs.python.org/3/library/shutil.html#shutil.move

While the temporary path *probably* supports `os.rename`, the cache directory could be on an esoteric filesystem that does not.

Reply:
If the cache filesystem does not support `os.rename`, then letting `shutil.move` fall back to a copy risks re-creating the concurrency issue I'm fixing here. I don't know if such esoteric file systems exist but perhaps we could address that someone complains? I can commit to following up on any issue raised about this post release.

Reply:
That's fair.

For reference this is the PR that changed it from `os.rename` to `shutil.move` : https://github.com/pypa/pip/pull/2749